binder/src/test/java/io/grpc/binder/BinderTransportSecurityTest.java

```diff
@@ -0,0 +1,235 @@
+        .setType(MethodDescriptor.MethodType.UNARY)
+        .setSampledToLocalTracing(true)
+        .build();
+  }
+
+  private static class SomeService extends LifecycleService {
+
+    private final IBinderReceiver binderReceiver = new IBinderReceiver();
+    private final ArrayBlockingQueue<SettableFuture<Status>> statusesToSet =
+        new ArrayBlockingQueue<>(128);
+    private Server server;
+    private HandlerThread handlerThread;
+    private Handler handler;
+
+    @Override
+    public void onCreate() {
+      super.onCreate();
+      handlerThread = new HandlerThread("test_handler_thread");
+      handlerThread.start();
+      handler = new Handler(handlerThread.getLooper());
+
+      MethodDescriptor<Empty, Empty> methodDesc = getMethodDescriptor();
+      ServerCallHandler<Empty, Empty> callHandler =
+          ServerCalls.asyncUnaryCall(
+              (req, respObserver) -> {
+                respObserver.onNext(req);
+                respObserver.onCompleted();
+              });
+      ServerMethodDefinition<Empty, Empty> methodDef =
+          ServerMethodDefinition.create(methodDesc, callHandler);
+      ServerServiceDefinition def = ServerServiceDefinition.builder(SERVICE_NAME)
+          .addMethod(methodDef)
+          .build();
+
+      server = BinderServerBuilder.forAddress(
+              AndroidComponentAddress.forContext(this),
+              binderReceiver)
+          .addService(def)
+          .securityPolicy(ServerSecurityPolicy.newBuilder()
+              .servicePolicy(SERVICE_NAME, new AsyncSecurityPolicy() {
+                @Override
+                ListenableFuture<Status> checkAuthorizationAsync(int uid) {
+                  return Futures.submitAsync(() -> {
+                    SettableFuture<Status> status = SettableFuture.create();
+                    statusesToSet.add(status);
+                    return status;
+                  }, getExecutor());
+                }
+              })
+              .build())
+          .executor(getExecutor())
+          .build();
+      try {
+        server.start();
+      } catch (IOException e) {
+        throw new IllegalStateException(e);
+      }
+
+      Application context = ApplicationProvider.getApplicationContext();
+      ComponentName componentName = new ComponentName(context, SomeService.class);
+      shadowOf(context)
+          .setComponentNameAndServiceForBindService(
+              componentName, checkNotNull(binderReceiver.get()));
+    }
+
+    /**
+     * Returns an {@link Executor} under which all of the gRPC computations run. The execution of
+     * any pending tasks on this executor can be triggered via {@link #idleLooper()}.
+     */
+    Executor getExecutor() {
+      return handler::post;
+    }
+
+    void idleLooper() {
+      shadowOf(handlerThread.getLooper()).idle();
+    }
+
+    void setSecurityPolicyStatusWhenReady(Status status) {
+      Uninterruptibles.takeUninterruptibly(statusesToSet).set(status);
```

Comment:
This also appears to block the main thread, in the common case where the server hasn't seen the request yet.

Reply:
That was intentional. At this point, the test has nothing else to do but wait until the gRPC's executor has ran, invoked the `AsyncSecurityPolicy` and placed a new `SettableFuture<Status>` into the queue.

This is meant to work somewhat like a [CyclicBarrier](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/CyclicBarrier.html): I want to have the test thread and the gRPC thread synchronized at the same point: after [this line](https://github.com/grpc/grpc-java/blob/fdd6f5001f354d5b57db976c4cbace8d64652452/binder/src/main/java/io/grpc/binder/internal/BinderTransportSecurity.java#L109). Thus, the test must wait until the queue has an entry as a signal that the implementation has generated the future.

Reply:
But how do you know that the gRPC executor hasn't enqueued something to run on the main thread and won't run the ServerCallHandler until that completes => Now you are deadlocked because your main thread won't service the event loop again until the ServerCallHandler runs.

It seems to me that you need to poll for this, not block, and idle() the single main looper between take() attempts.